### ISSUE TITLE
Unable to calculate cost of some charges

### ISSUE THREAD

**Issue by Chimpepe** [MattJeanes/TeslaMateAgile#68]
Hello Matt, 
since the changeover to winter time, I have had the problem that the prices are no longer calculated.
I have seen that there have been problems in the past. [https://github.com/MattJeanes/TeslaMateAgile/issues/41]

Do you have the chance to take a look at this?

Release v1.16.0

Thanks,
Dawid

```
teslamateagile-1  | 2024-11-07 21:48:03.189 dbug: TeslaMateAgile.PriceService[0]
teslamateagile-1  |       Updating prices
teslamateagile-1  | 2024-11-07 21:48:03.264 info: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Looking for finished charging processes with no cost set in the 'Home' geofence (id: 1)
teslamateagile-1  | 2024-11-07 21:48:03.345 info: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculating cost for charges 11/03/2024 10:04:56 UTC - 11/03/2024 13:00:16 UTC
teslamateagile-1  | 2024-11-07 21:48:03.501 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Retrieved 4 prices:
teslamateagile-1  | 2024-11-07 21:48:03.501 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/03/2024 09:00:00 UTC - 11/03/2024 10:00:00 UTC: 0.2679
teslamateagile-1  | 2024-11-07 21:48:03.501 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/03/2024 10:00:00 UTC - 11/03/2024 11:00:00 UTC: 0.2561
teslamateagile-1  | 2024-11-07 21:48:03.501 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/03/2024 11:00:00 UTC - 11/03/2024 12:00:00 UTC: 0.2464
teslamateagile-1  | 2024-11-07 21:48:03.501 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/03/2024 12:00:00 UTC - 11/03/2024 13:00:00 UTC: 0.2564
teslamateagile-1  | 2024-11-07 21:48:03.501 info: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Phase correction: 2 -> 3
teslamateagile-1  | 2024-11-07 21:48:03.513 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculated charge cost for 11/03/2024 10:00:00 UTC - 11/03/2024 11:00:00 UTC (unit cost: 0.2561, fee per kWh: 0): 2.530155393897083300453508 for 9.87956030416666653828 energy
teslamateagile-1  | 2024-11-07 21:48:03.525 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculated charge cost for 11/03/2024 11:00:00 UTC - 11/03/2024 12:00:00 UTC (unit cost: 0.2464, fee per kWh: 0): 2.691978432938666663210496 for 10.92523714666666665264 energy
teslamateagile-1  | 2024-11-07 21:48:03.537 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculated charge cost for 11/03/2024 12:00:00 UTC - 11/03/2024 13:00:00 UTC (unit cost: 0.2564, fee per kWh: 0): 2.786099471221333349428416 for 10.86622258666666672944 energy
teslamateagile-1  | 2024-11-07 21:48:03.537 fail: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Failed to calculate charging cost / energy for charging process 287
teslamateagile-1  |       System.Exception: Charge calculation failed, pricing calculated for 499 / 500, likely missing price data
teslamateagile-1  |          at TeslaMateAgile.PriceHelper.CalculateDynamicChargeCost(IEnumerable`1 charges, DateTimeOffset minDate, DateTimeOffset maxDate) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 151
teslamateagile-1  |          at TeslaMateAgile.PriceHelper.CalculateChargeCost(IEnumerable`1 charges) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 98
teslamateagile-1  |          at TeslaMateAgile.PriceHelper.Update() in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 75
teslamateagile-1  | 2024-11-07 21:48:03.538 info: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculating cost for charges 11/07/2024 10:35:28 UTC - 11/07/2024 12:00:05 UTC
teslamateagile-1  | 2024-11-07 21:48:03.606 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Retrieved 3 prices:
teslamateagile-1  | 2024-11-07 21:48:03.606 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/07/2024 09:00:00 UTC - 11/07/2024 10:00:00 UTC: 0.3609
teslamateagile-1  | 2024-11-07 21:48:03.606 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/07/2024 10:00:00 UTC - 11/07/2024 11:00:00 UTC: 0.3442
teslamateagile-1  | 2024-11-07 21:48:03.606 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       11/07/2024 11:00:00 UTC - 11/07/2024 12:00:00 UTC: 0.3445
teslamateagile-1  | 2024-11-07 21:48:03.606 info: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Phase correction: 2 -> 3
teslamateagile-1  | 2024-11-07 21:48:03.608 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculated charge cost for 11/07/2024 10:00:00 UTC - 11/07/2024 11:00:00 UTC (unit cost: 0.3442, fee per kWh: 0): 1.535331128878666432185006 for 4.46057852666666598543 energy
teslamateagile-1  | 2024-11-07 21:48:03.621 dbug: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Calculated charge cost for 11/07/2024 11:00:00 UTC - 11/07/2024 12:00:00 UTC (unit cost: 0.3445, fee per kWh: 0): 3.790141417660000047237840 for 11.00186188000000013712 energy
teslamateagile-1  | 2024-11-07 21:48:03.622 fail: TeslaMateAgile.PriceHelper[0]
teslamateagile-1  |       Failed to calculate charging cost / energy for charging process 288
teslamateagile-1  |       System.Exception: Charge calculation failed, pricing calculated for 247 / 248, likely missing price data
teslamateagile-1  |          at TeslaMateAgile.PriceHelper.CalculateDynamicChargeCost(IEnumerable`1 charges, DateTimeOffset minDate, DateTimeOffset maxDate) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 151
teslamateagile-1  |          at TeslaMateAgile.PriceHelper.CalculateChargeCost(IEnumerable`1 charges) in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 98
teslamateagile-1  |          at TeslaMateAgile.PriceHelper.Update() in /src/TeslaMateAgile/Helpers/PriceHelper.cs:line 75
teslamateagile-1  | 2024-11-07 21:48:03.636 dbug: TeslaMateAgile.PriceService[0]
teslamateagile-1  |       Price update complete
teslamateagile-1  | 2024-11-07 21:48:03.636 info: TeslaMateAgile.PriceService[0]
teslamateagile-1  |       Waiting 300 seconds until next update
```
[data-1731017330646.csv](https://github.com/user-attachments/files/17669974/data-1731017330646.csv)
[data-1731017347818.csv](https://github.com/user-attachments/files/17669976/data-1731017347818.csv)

**Comment by MattJeanes** [comment]
Hey, thank you for the detailed debug logs and CSVs - very useful! I'll also need your TeslaMateAgile configuration too if that's cool e.g. which provider you are using and the configuration for it (excluding any sensitive info of course!)

**Comment by Chimpepe** [comment]
```
teslamateagile:
    image: mattjeanes/teslamateagile:latest
    restart: always
    environment:
      - DATABASE_USER=teslamate
      - DATABASE_PASS=1234567890
      - DATABASE_NAME=teslamate
      - DATABASE_HOST=database
      - TeslaMate__UpdateIntervalSeconds=300
      - TeslaMate__GeofenceId=1
      - TeslaMate__EnergyProvider=Tibber
      - Tibber__AccessToken=1234567890
```

**Comment by MattJeanes** [comment]
Thank you! Looks like I wasn't correctly accounting for the end of the time window in certain edge cases in the Tibber provider, which hopefully the new preview version fixes.

Can you try out [v1.17.0-rc.4](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.17.0-rc.4) and let me know if that fixes it for you please?

```yaml
teslamateagile:
    image: mattjeanes/teslamateagile:v1.17.0-rc.4
```

**Comment by MattJeanes** [comment]
Hey @Chimpepe - did you get a chance to try out the new version? How's it going for you?

**Comment by Chimpepe** [comment]
Hey Matt, unfortunately not yet, have not charged due to the current high prices 😓

**Comment by MattJeanes** [comment]
No worries - keep me posted! Just want to make sure it's all working okay before I release the update properly 😄

**Comment by Chimpepe** [comment]
So, finally. Test is running. I'll let you know the result!

**Comment by Chimpepe** [comment]
Hey @MattJeanes ooks like it's missing again.

[data-1732454147050.csv](https://github.com/user-attachments/files/17893395/data-1732454147050.csv)
<img width="1322" alt="Bildschirmfoto 2024-11-24 um 14 16 46" src="https://github.com/user-attachments/assets/2f003d9d-d06c-4b05-b436-a7b3c715b239">

**Comment by Chimpepe** [comment]
Ahh, now I also have the costs. Seems like I had to wait until the started hour was over.

So looks like it's working.

**Comment by Chimpepe** [comment]
I will charge again tonight, then I can tell you in the morning if everything went well.

Thanks for your gorgeous support!

**Comment by Chimpepe** [comment]
Charging tonight also okay.

From my side all fine.

**Comment by MattJeanes** [comment]
Thank you!

This fix has now been released in version [`v1.17.0`](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.17.0) 🙂
